pkg/clientcert/cert_controller.go

```diff
@@ -0,0 +1,329 @@
+package clientcert
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"math/rand"
+	"reflect"
+	"time"
+
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+
+	certificates "k8s.io/api/certificates/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	certificatesinformers "k8s.io/client-go/informers/certificates/v1beta1"
+	corev1informers "k8s.io/client-go/informers/core/v1"
+	csrclient "k8s.io/client-go/kubernetes/typed/certificates/v1beta1"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+	certificateslisters "k8s.io/client-go/listers/certificates/v1beta1"
+	cache "k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/keyutil"
+	"k8s.io/klog/v2"
+)
+
+const (
+	// KubeconfigFile is the name of the kubeconfig file in kubeconfigSecret
+	KubeconfigFile = "kubeconfig"
+	// TLSKeyFile is the name of tls key file in kubeconfigSecret
+	TLSKeyFile = "tls.key"
+	// TLSCertFile is the name of the tls cert file in kubeconfigSecret
+	TLSCertFile = "tls.crt"
+
+	clusterNameAnnotation = "open-cluster-management.io/cluster-name"
+	ClusterNameFile       = "cluster-name"
+	AgentNameFile         = "agent-name"
+
+	irrelevantSecretKey = "irrelevantSecretKey"
+
+	ClusterNameLabel = "open-cluster-management.io/cluster-name"
+	AddonNameLabel   = "open-cluster-management.io/addon-name"
+	SignerNameLabel  = "open-cluster-management.io/signer-name"
+)
+
+// ControllerResyncInterval is exposed so that integration tests can crank up the constroller sync speed.
+var ControllerResyncInterval = 5 * time.Minute
+
+// CertificateManager abstracts behaviours which might be different during the creation/rotation of client
+// certification.
```

Comment:
I think you may add some doc about cluster registration and addon registration will implement the  interface.